Hindsight 2020 (#12)
image-rs/blog.image-rs.org: _posts/2021-02-10-hindsight-on-2020.md

supplementary information such as color spaces, comments, copyright,
orientation, and other parts of EXIF. This starts at concrete treatment of
extension chunks in the decoder libraries and goes all the way to design work
for integrating those in the `ImageDecoder` trait and the `Reader`.

* <https://github.com/image-rs/image-gif/issues/40>: Allow reading loop control extension (so that it can be written again)
* <https://github.com/image-rs/image-gif/issues/79>: The treatment of extensions
* <https://github.com/image-rs/image-png/issues/116>: Support for encoding metadata chunks
* <https://github.com/image-rs/image-png/issues/221>: Treatment of unknown chunks
* <https://github.com/image-rs/image/issues/258>: Need more portable image coordinates.
* <https://github.com/image-rs/image/issues/1371>: Orientation field not found after load image and save

## Limits during decoding

Loading an image from a remote server, only to find out that it decompresses to
several gigabytes of memory and totally grinds your system to a halt/crashes
your program isn't great. We'd like to ensure that the decoders all have memory
and/or runtime limits that they check and abide to. This also helps with

> fuzzing as it turns allows controlling for such use explicitly. Note that most

Reviewer: nit: turns -> in turn

Reviewer 2: 0x570D401AD99fF3f33CDdEdCcb4cfD29D3CDA33BB

contribute towards in 2021.

We also have a logo now:

<div style="background: center/contain no-repeat url('/assets/image-rs-logo.png'); width:100%; height:10em;"></div>

## The gif crate now 2x faster

The `gif` crate joins `png` in being on par with the relevant C implementation
in terms of performance!

These gains are thanks to a new LZW encoding/decoding library, [called
`weezl`][weezl]. It significantly improves on lzw crate in performance while
staying 100% safe, and also had an optional `no_std` mode. There is integration
with both `std` and `async` IO-interfaces, behind separate features. `weezl`,
written by the author of this post, is also maintained under the image-rs
organization that hosts a lot of format-specific image encoding/decoding
crates.

Since most of the time in decoding GIF files is spent in LZW decoding, the
gains from LZW have naturally translated to gif crate. TIFF files encoded with
LZW compression also benefit from this change. The latest versions of gif, tiff
and image all take advantage of this new library.

[weezl]: https://crates.io/crates/weezl

## Developments in 2020

Part of the stability work was the ability to upgrade decoder dependencies

> behind the scenes. This allowed you to silently benefit from the improved `gif`

Reviewer 3: ```suggestion
Part of the stability work was the ability to upgrade the crates with decoder implementations such as `gif` without breaking the API of the `image` crate.
```

decoder as detailed above and other similar fixes across other decoders and it
will also enable similar updates for `png` soon.

This is directly reflected in dependencies and downloads. While previously
there was a long tail, the dependencies and downloads are now [dominantly for
recent versions](https://lib.rs/crates/image/rev) and thus quicker, more
secure, and more correct. (If you're still using `0.18` or `0.22`, you should
_really_ consider updating). We've also more than doubled our total downloads

> over the last year!

Reviewer 3: More secure? There were security bugs fixed in non-semver-compatible releases? Uh oh. 

The part about "long tail" is also a bit hard to grok.

I'd rewrite the entire paragraph like this:

```suggestion
This is directly reflected in dependencies and downloads. We have fewer downloads of older versions and fewer crates depending on an outdated version than ever! If you're still using `0.18` or `0.22`, you should _really_ consider updating. We've also more than doubled our total downloads over the last year!
```

Author: Yeah, the one instance I'm (not directly) talking about here was `0.21.3` to `0.22` which dropped one generic interface that would have needed a `bytemuck::Pod` constraint to be sound, which didn't exist at the time. It's not in 2020 though so I'll probably go with dropping the sentence.

> All of this didn't stop progress. Take the `png` and `tiff` decoders that have

Reviewer 3: ```suggestion
Despite the API stability, we've made a ton of progress! The the `png` and `tiff` crates have
```

> seen a significant increase in coverage of features. Add the fact that `png`

Reviewer 3: "a significant increase in coverage of features." is not impressive. We need to brag! What are those features? How do they make the world a better place?

Reviewer 3: To make the achievements even more impressive, spread them across categories instead of a single paragraph with everything lumped together. Rough outline to provide an example:

**Performance**

gif and png are on par with C, improvements to jpeg but it's hard to compete with SIMD assembly; huffman decoding is slowing things down, it's over here, please help us optimize it.

**Features**

<list and description of the newly added features goes here>

**Stability**

<the image stable API goes here>

**Compatibility**

<bragging about running hundreds of thousands of images through each decoder goes here>

> and `gif` are now quite competitive with their C counterparts. Also, the

Reviewer 3: ```suggestion
seen a significant increase in coverage of features, and their performance is now quite competitive with their C counterparts. Also, the
```

`jpeg-decoder` crate that has fixed quite a lot of outstanding bugs in the
decoding process; The newly added bindings for `avif` decoders and encoders;

> And not to forget improved and copyless image operations.

Reviewer 3: Improved how?

> 

Reviewer 3: I suggest adding another paragraph on testing on real-world data. That we've scraped _hundreds of thousands_ of images off the web and ran them through the decoders to see what happens, and that they mostly pass now!

> There are still many outstanding issues and unimplemented improvements. From

Reviewer 3: There should be a header before this paragraph, since we're talking about a completely different thing now.

Depending on the blog CSS, it might be a good idea to bump the other headings to level 3.

the point of view of a maintainer, some of the most pressing ones are listed in
the next sections, sorted roughly by topic.

## Fallbacks and Error logs

The decoders are strict, in many ways. There exist a few fallbacks here and

> there, but not enough. The justification for this is that being too lenient in

Reviewer 3: Not enough for what?

```suggestion
there, but we still occasionally encounter a technically invalid image that is successfully decoder by other tools. The justification for this is that being too lenient in
```

accepted formats leads to an ossification of formats where particular
non-standard behavior by the most popular encoders is de-facto standardized
without appearing in any specification. For many older formats that ship has,
however, long sailed. We should thus accept the popular interpretation (e.g.
ImageMagick) of such extensions.

The issues goes a bit farther than this. It would be great to have a mode where
such details and nit-picks are remarked upon but decoding continues. Meanwhile

> we do not want to dump them to `stderr`.

Reviewer 3: ```suggestion
The issue goes a bit farther than this. It would be great to have a mode where
such details and nit-picks are reported, but decoding continues. But we don't want to
just dump them to `stderr`, like some C libraries do.
```

* <https://github.com/image-rs/image/issues/1371>: Orientation field not found after load image and save

## Limits during decoding

Loading an image from a remote server, only to find out that it decompresses to
several gigabytes of memory and totally grinds your system to a halt/crashes
your program isn't great. We'd like to ensure that the decoders all have memory
and/or runtime limits that they check and abide to. This also helps with
fuzzing as it turns allows controlling for such use explicitly. Note that most
of the core libraries (png, gif, tiff) all have their own form of limits but
notably `jpeg-decoder` does not. Also, there is no common interface to control
them in `image` and to restrict decoding to such parsers that can enforce them.

* <https://github.com/image-rs/image/issues/938>: Library wide memory limits
* <https://github.com/image-rs/image/issues/1052>: Large CPU and memory consumption on decoding a crafted GIF file
* <https://github.com/image-rs/image/issues/1139>: Excessively slow parsing of certain files
* <https://github.com/image-rs/jpeg-decoder/issues/133>: Excessive memory allocation for small inputs

> * <https://github.com/image-rs/jpeg-decoder/issues/134>: Slow parsing for small inputs

Reviewer 3: Needs an issue for JPEG not having a memory limit.

Author: Issue 133 is about adding limits—at least as a proposal in comments—so I've clarified its title.

## Code style and Documentation

Code changes and gets old, sometimes quite a bit and very quickly. This can
manifest as the coding style of one part of the library being inconsistent with
the rest, leading to new contributors unwittingly basing their contributions on
outdated styles. This is a bit of problem for both parties: It makes reviews

> harder and doesn't teach the right values to potentially new programmers.

Reviewer 3: I understand "coding style" as "tabs vs spaces", and that's pretty handily solved by `rustfmt`. So this does not make a great deal of sense to me.

If there's no clear way to salvage this paragraph, I suggest simply dropping it.

Author: No, this isn't about `rustfmt`. This is about such things as `vec.extend(slice.iter().cloned())` that were completely superseeded, or about workarounds around borrow checker v1.

Author: I've rewritten it, please take a look.

Reviewer 3: Looks good! Marking resolved

`ImageBuffer` is showing a few cracks in its design but replacing it directly
is also not quite feasible. It has chosen _one_ layout and encodes this within
its own type. Both of these aspects have proven to be too inflexible.

Using `Vec<T>` for representing pixels is not very efficient if one wants to
support operations that change the sample type. This is due to the fact that it
relies on the standard libraries use of allocators and the memory allocated for
a vector is tied to the _exact_ layout of the sample type. There is a concept
library to work around this restriction by storing everything as a highly
aligned byte buffer, which makes it unnecessary to track the sample type as a
type parameter to the buffer type itself. It's a work in progress found here:

<https://github.com/image-rs/canvas>

The flexible layout makes it seem like a reasonable goal to create integration
with the various texture layouts found in graphics crates such as [`wgpu`],
[`ash`], [`gtk`], the Linux DRM module, and [`image`], and enable some
interoperability between them.

[`wgpu`]: https://crates.io/crates/wgpu
[`ash`]: https://crates.io/crates/ash
[`gtk`]: https://crates.io/crates/gtk

> [`image`]: https://crates.io/crates/image

Reviewer 3: Needs a conclusion / call to action. Right now the article kind of fizzles out at the end. Try ending on a high point: Image encoding, decoding and manipulation in safe Rust is better than ever before! If you want to help, start here or here!

with both `std` and `async` IO-interfaces, behind separate features. `weezl`,
written by the author of this post, is also maintained under the image-rs
organization.

Since most of the time in decoding GIF files is spent in LZW decoding, the
gains from LZW have naturally translated to gif crate. TIFF files encoded with
LZW compression also benefit from this change. The total decoding throughput
quadrupled in some instances! The latest versions of gif, tiff and image all
take advantage of this new library.

[weezl]: https://crates.io/crates/weezl

The `jpeg-decoder` crate has also been sped up significantly. Idiomatic
iterator based loops of Rust allow the LLVM compiler to aggressively use auto
vectorization. That is, instead of performing a transformation on single
elements at a time this optimization technique automatically replaces code
operating on elements sequentially (in this case, pixels) with special SIMD
instructions that perform them on multiple elements at a time, with large
speedups of 400% or more. While not yet on-par with handwritten assembly this
can get very close with the right code structure. Additionally, a number of

> index checks have been identified. Together, the whole decoding process takes

Reviewer 3: > Additionally, a number of index checks have been identified.

Not super clear. Try: "Iterator-based approach also eliminated some needless bounds checks, which improved performance without sacrificing safety."

Reviewer 3: Also, you forgot to include increased parallelism - IDCT is now has an independent thread for each component rather than a single worker thread: https://github.com/image-rs/jpeg-decoder/pull/168
This speeds up decoding color images (or it would, if the huffman decoding would not hold up the entire thing)

library to work around this restriction by storing everything as a highly
aligned byte buffer, which makes it unnecessary to track the sample type as a
type parameter to the buffer type itself. It's a work in progress found here:

<https://github.com/image-rs/canvas>

The flexible layout makes it seem like a reasonable goal to create integration
with the various texture layouts found in graphics crates such as [`wgpu`],
[`ash`], [`gtk`], the Linux DRM module, and [`image`], and enable some
interoperability between them.

[`wgpu`]: https://crates.io/crates/wgpu
[`ash`]: https://crates.io/crates/ash
[`gtk`]: https://crates.io/crates/gtk
[`image`]: https://crates.io/crates/image

## A call for action

The progress report hopefully inspired as much sense of progress as I had
writing this post, and conveyed the vision of where we want to go. You can help
us get there! No matter if you simply want to learn about image formats, if you
want to contribute to the ecosystem, if you your employer needs a specific
feature, or if you're here for other reasons entirely: Grab an issue,
investigate, open a PR, and we'll help you get it upstream. Let's make 2021 at

> least as successful for `image-rs` as last year.

Reviewer 3: :+1:

Since most of the time in decoding GIF files is spent in LZW decoding, the
gains from LZW have naturally translated to gif crate. TIFF files encoded with
LZW compression also benefit from this change. The total decoding throughput
quadrupled in some instances! The latest versions of gif, tiff and image all
take advantage of this new library.

[weezl]: https://crates.io/crates/weezl

The `jpeg-decoder` crate has also been sped up significantly. Idiomatic
iterator based loops of Rust allow the LLVM compiler to aggressively use auto
vectorization. That is, instead of performing a transformation on single
elements at a time this optimization technique automatically replaces code
operating on elements sequentially (in this case, pixels) with special SIMD
instructions that perform them on multiple elements at a time, with large
speedups of 400% or more. While not yet on-par with handwritten assembly this
can get very close with the right code structure.

Additionally, a number of needless bounds checks have been identified. It was
possible to remove them with an iterator-based approach, which improved
performance without sacrificing safety.  Together, the whole decoding process

> takes 20%-30% less processor time. At last, the decoder learned to make better

Reviewer 3: ```suggestion
takes 20%-30% less processor time. Finally, the decoder now makes better
```

recent versions](https://lib.rs/crates/image/rev) and thus quicker, using less
memory, more correct and with less panics. (If you're still using `0.18` or
`0.22`, you should _really_ consider updating). We've also more than doubled
our total downloads over the last year! Around 1_000 (one thousand!) crates now

> depend one at least one of the crates under our organization. That's mind

Reviewer 4: "depend _one_" should be "depend _on_".

boggling.

Despite the API stability, we've made a ton of progress!

### Formats

The list of supported formats has expanded greatly. The formats `farbfeld`,
`tga`, `avif` and several DDS subtypes are now recognized or can be written.
The existing formats also cover more of the specification.

Let's start with the basics. The `png` library now supports decoding of
individual APNG frames. It also now treats all allowed pixel formats properly,
including sixteen bit depth. This all makes it pass the extended `pngsuite`
with flying colors. It also gained the ability to encode indexed data directly.

The `tiff` crate has also seen a lot of attention. Throughout the year various
contributors added BigTIFF and basic GeoTIFF support, have expanded the
supported sample formats to 32-bit and 64-bit color depth, and overall
increased the amount of recognized tags. It also decodes embedded modern jpeg
image data.  Nevertheless, one particularly sore point remains: `tiff` does not
yet support Tile Offsets which would be extremely valuable for access to large
data from medical and geodetic data. We'd be delighted to see contributions for
this feature.

> The `jpeg-decoder` crate now correctly decoders all progressive images. It now

Reviewer 4: "The `jpeg-decoder` crate now correctly _decoders_" should be "The `jpeg-decoder` crate now correctly _decodes_".
